File: server/core/cipher.go
package core

import (
	"fmt"
	"math/rand"
	"strconv"
	"strings"
	"time"
)

const numRounds = 64
const BlockSize = 8
const delta = 0x9E3779B9

type KeySizeError int
type Cipher struct {
	table [64]uint32
}

func (c *Cipher) BlockSize() int          { return BlockSize }
func (c *Cipher) Encrypt(dst, src []byte) { encryptBlock(c, dst, src) }
func (c *Cipher) Decrypt(dst, src []byte) { decryptBlock(c, dst, src) }

func CryptCreateKey(length int) []byte {

	rand.Seed(time.Now().UnixNano())
	str := make([]byte, length)

	for i := range str {
		str[i] = Characters[rand.Intn(len(Characters))]
	}

	return str
}

func CreateHashMacro(str string) string {

	macro := strings.ToUpper(strings.TrimRight(str, "\n"))
	hash := GetHashFromString(str)

	return fmt.Sprintf("#define %s 0x%x", strings.Split(macro, ".")[0], hash)
}

func GetHashFromString(str string) uint32 {
	var (
		name    string
		sum32   = uint32(2166136261)
		prime32 = uint32(16777619)
	)

	lower := strings.ToLower(str)

	if strings.HasSuffix(lower, ".dll") {
		name = string(EncodeUTF16(lower))
	} else {
		name = lower
	}

	hash := sum32
	for i := 0; i < len(name); i++ {
		hash ^= uint32(name[i])
		hash *= prime32
	}

	return hash
}

func NewCipher(key []byte) (*Cipher, error) {

	k := len(key)

	switch k {
	default:
		return nil, KeySizeError(k)
	case 16:
		break
	}

	c := new(Cipher)
	initCipher(c, key)

	return c, nil
}

func XteaDivide(data []byte) [][]byte {
	var (
		sections    [][]byte
		sectionSize = 8
	)

	for i := 0; i < len(data); i += sectionSize {
		end := i + sectionSize
		sections = append(sections, data[i:end])
	}

	return sections
}

func CryptXtea(config, key []byte, encrypt bool) ([]byte, error) {
	var (
		out    []byte
		err    error
		cipher *Cipher
	)

	if cipher, err = NewCipher(key); err != nil {
		return nil, err
	}

	sections := XteaDivide(config)
	out = make([]byte, 0)

	for _, section := range sections {
		buf := make([]byte, 8)

		if encrypt {
			cipher.Encrypt(buf, section)
		} else {
			cipher.Decrypt(buf, section)
		}

		out = append(out, buf...)
	}

	return out, nil
}

func blockToUint32(src []byte) (uint32, uint32) {
	r0 := uint32(src[0])<<24 | uint32(src[1])<<16 | uint32(src[2])<<8 | uint32(src[3])
	r1 := uint32(src[4])<<24 | uint32(src[5])<<16 | uint32(src[6])<<8 | uint32(src[7])
	return r0, r1
}

func uint32ToBlock(v0, v1 uint32, dst []byte) {
	dst[0] = byte(v0 >> 24)
	dst[1] = byte(v0 >> 16)
	dst[2] = byte(v0 >> 8)
	dst[3] = byte(v0)
	dst[4] = byte(v1 >> 24)
	dst[5] = byte(v1 >> 16)
	dst[6] = byte(v1 >> 8)
	dst[7] = byte(v1 >> 0)
}

func encryptBlock(c *Cipher, dst, src []byte) {

	v0, v1 := blockToUint32(src)

	for i := 0; i < numRounds; {
		v0 += ((v1<<4 ^ v1>>5) + v1) ^ c.table[i]
		i++

		v1 += ((v0<<4 ^ v0>>5) + v0) ^ c.table[i]
		i++
	}

	uint32ToBlock(v0, v1, dst)
}

func decryptBlock(c *Cipher, dst, src []byte) {

	v0, v1 := blockToUint32(src)

	for i := numRounds; i > 0; {
		i--
		v1 -= ((v0<<4 ^ v0>>5) + v0) ^ c.table[i]

		i--
		v0 -= ((v1<<4 ^ v1>>5) + v1) ^ c.table[i]
	}

	uint32ToBlock(v0, v1, dst)
}

func (k KeySizeError) Error() string {
	return "invalid key size " + strconv.Itoa(int(k))
}

func initCipher(c *Cipher, key []byte) {
	var (
		k   [4]uint32
		sum uint32
	)

	for i := 0; i < len(k); i++ {
		j := i << 2 // Multiply by 4
		k[i] = uint32(key[j+0])<<24 | uint32(key[j+1])<<16 | uint32(key[j+2])<<8 | uint32(key[j+3])
	}

	for i := 0; i < numRounds; {
		c.table[i] = sum + k[sum&3]
		i++
		sum += delta
		c.table[i] = sum + k[(sum>>11)&3]
		i++
	}
}

File: rust_server/src/server/cipher.rs
use std::time::{SystemTime, UNIX_EPOCH};
use rand::{Rng, SeedableRng};
use rand::rngs::StdRng;
use std::error::Error;
use std::convert::TryInto;
use crate::server::error::KeySizeError;

const NUM_ROUNDS:   usize = 64;
const BLOCK_SIZE:   usize = 8;
const DELTA:        u32 = 0x9E3779B9;
const FNV_OFFSET:   u32 = 2166136261;
const FNV_PRIME:    u32 = 16777619;
const CHARACTERS:   &[u8] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";


struct Cipher {
    table: [u32; NUM_ROUNDS],
}

impl Cipher {
    fn block_size(&self) -> usize {
        BLOCK_SIZE
    }

    fn encrypt(&self, dst: &mut [u8], src: &[u8]) {
        encrypt_block(self, dst, src);
    }

    fn decrypt(&self, dst: &mut [u8], src: &[u8]) {
        decrypt_block(self, dst, src);
    }
}

pub fn crypt_create_key(length: usize) -> Vec<u8> {
    let seed = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_nanos();
    let mut rng = StdRng::seed_from_u64(seed as u64);
    let mut key = vec![0u8; length];

    for i in 0..length {
        key[i] = CHARACTERS[rng.gen_range(0..CHARACTERS.len())];
    }

    key
}

fn get_hash_from_string(s: &str, is_unicode: bool) -> u32 {
    let mut hash = FNV_OFFSET;
    let length = if is_unicode { s.len() - 2 } else { s.len() };
    let offset = if is_unicode { 2 } else { 1 };

    for i in (0..length).step_by(offset) {
        hash ^= s.as_bytes()[i] as u32;
        hash = hash.wrapping_mul(FNV_PRIME);
    }

    hash
}

fn create_hash_macro(s: &str) -> String {
    let macro_name  = s.to_uppercase().trim_end().to_string();
    let lower       = s.to_lowercase();
    let (name, is_unicode) = if lower.ends_with(".dll") {
        (encode_utf16(&lower), true)
    } else {
        (lower.into_bytes(), false)
    };

    format!(
        "#define {} 0x{:x}",
        macro_name.split('.').next().unwrap(),
        get_hash_from_string(&String::from_utf8_lossy(&name), is_unicode)
    )
}

fn new_cipher(key: &[u8]) -> Result<Cipher, KeySizeError> {
    if key.len() != 16 {
        return Err(KeySizeError(key.len()));
    }
    let mut cipher = Cipher {
        table: [0u32; NUM_ROUNDS]
    };

    init_cipher(&mut cipher, key);
    Ok(cipher)
}

fn xtea_divide(data: &[u8]) -> Vec<&[u8]> {
    data.chunks(BLOCK_SIZE).collect()
}

pub fn crypt_xtea(config: &[u8], key: &[u8], encrypt: bool) -> Result<Vec<u8>, Box<dyn Error>> {
    let cipher      = new_cipher(key)?;
    let sections    = xtea_divide(config);
    let mut out     = Vec::with_capacity(config.len());

    for section in sections {
        let mut buf = [0u8; BLOCK_SIZE];

        if encrypt  {
            cipher.encrypt(&mut buf, section);
        } else {
            cipher.decrypt(&mut buf, section);
        }

        out.extend_from_slice(&buf);
    }

    Ok(out)
}

fn block_to_u32(src: &[u8]) -> (u32, u32) {
    let r0 = u32::from_be_bytes(src[0..4].try_into().unwrap());
    let r1 = u32::from_be_bytes(src[4..8].try_into().unwrap());
    (r0, r1)
}

fn u32_to_block(v0: u32, v1: u32, dst: &mut [u8]) {
    dst[0..4].copy_from_slice(&v0.to_be_bytes());
    dst[4..8].copy_from_slice(&v1.to_be_bytes());
}

fn encrypt_block(c: &Cipher, dst: &mut [u8], src: &[u8]) {
    let (mut v0, mut v1) = block_to_u32(src);

    for i in 0..NUM_ROUNDS {
        v0 = v0.wrapping_add(((v1 << 4) ^ (v1 >> 5)).wrapping_add(v1) ^ c.table[i]);
        v1 = v1.wrapping_add(((v0 << 4) ^ (v0 >> 5)).wrapping_add(v0) ^ c.table[i]);
    }

    u32_to_block(v0, v1, dst);
}

fn decrypt_block(c: &Cipher, dst: &mut [u8], src: &[u8]) {
    let (mut v0, mut v1) = block_to_u32(src);

    for i in (0..NUM_ROUNDS).rev() {
        v1 = v1.wrapping_sub(((v0 << 4) ^ (v0 >> 5)).wrapping_add(v0) ^ c.table[i]);
        v0 = v0.wrapping_sub(((v1 << 4) ^ (v1 >> 5)).wrapping_add(v1) ^ c.table[i]);
    }

    u32_to_block(v0, v1, dst);
}

fn init_cipher(c: &mut Cipher, key: &[u8]) {
    let mut k = [0u32; 4];
    let mut sum = 0u32;

    for (i, k_val) in k.iter_mut().enumerate() {
        let j = i * 4;
        *k_val = u32::from_be_bytes(key[j..j + 4].try_into().unwrap());
    }

    for i in 0..NUM_ROUNDS {
        c.table[i] = sum.wrapping_add(k[(sum & 3) as usize]);
        sum = sum.wrapping_add(DELTA);
        c.table[i] = sum.wrapping_add(k[((sum >> 11) & 3) as usize]);
    }
}

fn encode_utf16(s: &str) -> Vec<u8> {
    s.encode_utf16().flat_map(|c| c.to_be_bytes()).collect()
}